![https://raw.githubusercontent.com/marciolws/Curso_EBAC_Cientista_de_Dados/refs/heads/main/EBAC-media-utils/logo/ebac_logo-data_science.png](https://raw.githubusercontent.com/marciolws/Curso_EBAC_Cientista_de_Dados/refs/heads/main/EBAC-media-utils/logo/ebac_logo-data_science.png)

---

<!-- # **Profissão: Cientista de Dados** -->
### **Módulo 27** | PCA | Exercício I

**Aluno:** [Marcio da Silva](https://www.linkedin.com/in/marcio-d-silva/)<br>
**Data:** 09 de outubro de 2024.

---


# PCA - Tarefa 01: *HAR* com PCA

Vamos trabalhar com a base da demonstração feita em aula, mas vamos explorar um pouco melhor como é o desempenho da árvore variando o número de componentes principais.

In [1]:
import pandas as pd

from sklearn.tree import DecisionTreeClassifier

from sklearn.decomposition import PCA
from sklearn.metrics import accuracy_score
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV


In [4]:
filename_features = "UCI HAR Dataset/features.txt"
filename_labels = "UCI HAR Dataset/activity_labels.txt"

filename_subtrain = "/Users/marciolewis/Library/Mobile Documents/com~apple~CloudDocs/EBAC_EXERCICIOS/Módulo 27 - PCA/UCI HAR Dataset/train/subject_train.txt"
filename_xtrain = "/Users/marciolewis/Library/Mobile Documents/com~apple~CloudDocs/EBAC_EXERCICIOS/Módulo 27 - PCA/UCI HAR Dataset/train/X_train.txt"
filename_ytrain = "/Users/marciolewis/Library/Mobile Documents/com~apple~CloudDocs/EBAC_EXERCICIOS/Módulo 27 - PCA/UCI HAR Dataset/train/y_train.txt"

filename_subtest = "/Users/marciolewis/Library/Mobile Documents/com~apple~CloudDocs/EBAC_EXERCICIOS/Módulo 27 - PCA/UCI HAR Dataset/test/subject_test.txt"
ffilename_xtest = "/Users/marciolewis/Library/Mobile Documents/com~apple~CloudDocs/EBAC_EXERCICIOS/Módulo 27 - PCA/UCI HAR Dataset/test/X_test.txt"
filename_ytest = "/Users/marciolewis/Library/Mobile Documents/com~apple~CloudDocs/EBAC_EXERCICIOS/Módulo 27 - PCA/UCI HAR Dataset/test/y_test.txt"

features = pd.read_csv(filename_features, header=None, names=['nome_var'], sep="#").squeeze('columns')
labels = pd.read_csv(filename_labels, delim_whitespace=True, header=None, names=['cod_label', 'label'])

subject_train = pd.read_csv(filename_subtrain, header=None, names=['subject_id'], sep="#").squeeze('columns')
X_train = pd.read_csv(filename_xtrain, delim_whitespace=True, header=None, names=features.tolist())
y_train = pd.read_csv(filename_ytrain, header=None, names=['cod_label'])

subject_test = pd.read_csv(filename_subtest, header=None, names=['subject_id'], sep="#").squeeze('columns')
X_test = pd.read_csv(ffilename_xtest, delim_whitespace=True, header=None, names=features.tolist())
y_test = pd.read_csv(filename_ytest, header=None, names=['cod_label'])

/var/folders/0w/j12vvsfd1191gn6cm0rsqyn40000gn/T/ipykernel_5952/35656892.py:13: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\s+'`` instead
  labels = pd.read_csv(filename_labels, delim_whitespace=True, header=None, names=['cod_label', 'label'])
/var/folders/0w/j12vvsfd1191gn6cm0rsqyn40000gn/T/ipykernel_5952/35656892.py:16: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\s+'`` instead
  X_train = pd.read_csv(filename_xtrain, delim_whitespace=True, header=None, names=features.tolist())
/var/folders/0w/j12vvsfd1191gn6cm0rsqyn40000gn/T/ipykernel_5952/35656892.py:20: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\s+'`` instead
  X_test = pd.read_csv(ffilename_xtest, delim_whitespace=True, header=None, names=features.tolist())


## Árvore de decisão

Rode uma árvore de decisão com todas as variáveis, utilizando o ```ccp_alpha=0.001```. Avalie a acurácia nas bases de treinamento e teste. Avalie o tempo de processamento.

In [9]:
%%time
#Medição do tempo de processamento da arvore.

#Criando o calssificador da arvore de decisão com o ccp_aplha=0.001.
clf = DecisionTreeClassifier(ccp_alpha=0.001)

#Treinando o classificador com os dados de treino.
clf.fit(X_train, y_train)

#Avaliação da acurácia nos dados de treinamento.
print(f'A acurácia na base de treinamento foi de: {clf.score(X_train, y_train)}')
#Avaliação da acurácia nos dados de teste.
print(f'A acurácia na base de teste foi de: {clf.score(X_test, y_test)}')


A acurácia na base de treinamento foi de: 0.9757889009793254
A acurácia na base de teste foi de: 0.8815744825246012
CPU times: user 4.72 s, sys: 39.1 ms, total: 4.76 s
Wall time: 4.95 s


## Árvore com PCA

Faça uma análise de componemtes principais das variáveis originais. Utilize apenas uma componente. Faça uma árvore de decisão com esta componente como variável explicativa.

- Avalie a acurácia nas bases de treinamento e teste
- Avalie o tempo de processamento

In [24]:
%%time

#Determinando o PCA com 1 comoponente para os dados de treinamento.
prcomp = PCA(n_components=1).fit(X_train)

#Transformando os dados de treinamento e teste com os componentes encontrados pelo PCA.
pc_treino = prcomp.transform(X_train)
pc_teste = prcomp.transform(X_test)

print(f'A dimenssão da base de dados de treinamento foi de: {pc_treino.shape}')
print(f'A dimenssão da base de dados de teste foi de: {pc_teste.shape}')

clf = DecisionTreeClassifier(ccp_alpha=0.001)
clf.fit(pc_treino, y_train)

print(f'A dimenssão da base de dados de treinamento foi de: {clf.score(pc_treino, y_train)}')
print(f'A dimenssão da base de dados de teste foi de: {clf.score(pc_teste, y_test)}\n')

A dimenssão da base de dados de treinamento foi de: (7352, 1)
A dimenssão da base de dados de teste foi de: (2947, 1)
A dimenssão da base de dados de treinamento foi de: 0.499727965179543
A dimenssão da base de dados de teste foi de: 0.45707499151679676

CPU times: user 279 ms, sys: 25.3 ms, total: 305 ms
Wall time: 265 ms


## Testando o número de componentes

Com base no código acima, teste a árvore de classificação com pelo menos as seguintes possibilidades de quantidades de componentes: ```[1, 2, 5, 10, 50]```. Avalie para cada uma delas:

- Acurácia nas bases de treino e teste
- Tempo de processamento


In [28]:
%%time

componentes = [1, 2, 5, 10, 50]

# Loop sobre os diferentes números de componentes
for n in componentes:
    # Executa o PCA com o número de componentes atual
    prcomp = PCA(n_components=n).fit(X_train)

    # Transforma os dados de treinamento e teste nos componentes principais
    pc_treino = prcomp.transform(X_train)
    pc_teste  = prcomp.transform(X_test)

    # Imprime as dimensões dos dados transformados
    print(f'Dimensões da base de treinamento: {pc_treino.shape}')
    print(f'Dimensões da base de teste: {pc_teste.shape}')

    # Cria e treina um classificador de árvore de decisão
    clf = DecisionTreeClassifier(ccp_alpha=0.001)
    clf.fit(pc_treino, y_train)

    # Avalia a acurácia na base de treinamento e teste
    print(f'Acurácia na base de treinamento: {clf.score(pc_treino, y_train)}')
    print(f'Acurácia na base de teste: {clf.score(pc_teste, y_test)}\n')

Dimensões da base de treinamento: (7352, 1)
Dimensões da base de teste: (2947, 1)
Acurácia na base de treinamento: 0.499727965179543
Acurácia na base de teste: 0.45707499151679676

Dimensões da base de treinamento: (7352, 2)
Dimensões da base de teste: (2947, 2)
Acurácia na base de treinamento: 0.6127584330794341
Acurácia na base de teste: 0.5846623685103495

Dimensões da base de treinamento: (7352, 5)
Dimensões da base de teste: (2947, 5)
Acurácia na base de treinamento: 0.8460282916213275
Acurácia na base de teste: 0.7885985748218527

Dimensões da base de treinamento: (7352, 10)
Dimensões da base de teste: (2947, 10)
Acurácia na base de treinamento: 0.8926822633297062
Acurácia na base de teste: 0.8242280285035629

Dimensões da base de treinamento: (7352, 50)
Dimensões da base de teste: (2947, 50)
Acurácia na base de treinamento: 0.919341675734494
Acurácia na base de teste: 0.822870715982355

CPU times: user 2.12 s, sys: 149 ms, total: 2.27 s
Wall time: 1.44 s


## Conclua

- O que aconteceu com a acurácia?
- O que aconteceu com o tempo de processamento?

> Com a redução da dimensionalidade utilizando a técnica de Análise de Componentes Principais (PCA), podemos observar que a principal mudança ocorre em relação ao tempo de processamento. A árvore executada com o conjunto de dados completo contendo todas as 561 variáveis, levou aproximadamente 4,9 segundos para ser executada, resultando em uma acurácia de 88% na base de teste. No entanto, ao utilizar o PCA com apenas um componente, o tempo de processamento foi reduzido para pouco menos de 300 ms, porém a acurácia na base de testes foi de 45%.

> Aumentando o número de componentes, foi possível aumentar gradualmente a acurácia na base de testes. Com 50 componentes, foi atingida uma acurácia de até 83%, com um tempo de processamento de aproximadamente 2 segundos para executar todas as 5 combinações de números diferentes de componentes. Portanto, podemos concluir que é possível obter uma acurácia semelhante à árvore com todas as variáveis do conjunto de dados, porém com um tempo de processamento reduzido em menos da metade. É importante ressaltar que os componentes são formados a partir da combinação linear das variáveis originais.